implementation/src/main/java/io/smallrye/config/ConfigSourceMap.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.smallrye.config;
+
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import org.eclipse.microprofile.config.spi.ConfigSource;
+
+/**
+ * A {@link Map Map&lt;String, String>} which is backed by a {@link ConfigSource}.
+ * This should not be used to implement {@link ConfigSource#getProperties()} on {@code ConfigSource}
+ * instances which do not override {@code getPropertyNames()}, as this will result in infinite recursion.
+ *
+ * @implNote The key set of the map is the result of calling {@link ConfigSource#getPropertyNames()}; the rest
+ * of the map operations are derived from this method and {@link ConfigSource#getValue(String)}.
+ * The values collection and entry set are instantiated lazily and cached.
+ * The implementation attempts to make no assumptions about the efficiency of the backing implementation and
+ * prefers the most direct access possible.
+ * <p>
+ * Some mutate operations are supported (specifically removals), where they can be performed on the backing key set, if
```

Comment:
that's not specified in the spec (although it should be) but my interpretation of `getPropertyNames()` is that it should return an immutable set of names.
In general, `ConfigSource` should not be used to modify/update configuration. If some of these methods provide ambiguity about it, it should be properly specified in the spec or in the API.

I'll open issues in `microprofile-config` to clarify these use cases

Reply:
The API says:  https://github.com/eclipse/microprofile-config/blob/4611c9433675e34ee73bd1eba026e01631582772/api/src/main/java/org/eclipse/microprofile/config/spi/ConfigSource.java#L39

Reply:
there is also https://github.com/eclipse/microprofile-config/issues/333 where it is discussed